src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+        const char *upath, flux_future_t **f)
+{
+    printf("Committing data to KVS\n");
+    *f = flux_kvs_commit(ctx->h, ctx->kvs_namespace,
+            0, txn);
+    if (*f == NULL)
+    {
+        DYAD_LOG_ERR(ctx, "Could not commit transaction to Flux KVS\n");
+        return DYAD_BADCOMMIT;
+    }
+    printf("Wait for commit to complete\n");
+    flux_future_wait_for(*f, -1.0);
+    printf("Destroy future for KVS commit\n");
+    flux_future_destroy(*f);
+    *f = NULL;
+    printf("Destroy transaction on KVS\n");
+    flux_kvs_txn_destroy(txn);
+    return DYAD_OK;
+}
+
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int publish_via_flux(dyad_ctx_t* ctx, const char *upath)
+{
+    printf("Initial publication setup\n");
+    const char *prod_managed_path = ctx->prod_managed_path;
+    flux_kvs_txn_t *txn = NULL;
+    flux_future_t *f = NULL;
+    const size_t topic_len = PATH_MAX;
+    char topic[topic_len+1];
+    printf("Creating KVS key\n");
+    memset(topic, '\0', topic_len+1);
+    gen_path_key(
+            upath,
+            topic,
+            topic_len,
+            ctx->key_depth,
+            ctx->key_bins
+            );
+    if (ctx->h == NULL)
+    {
+        fprintf (stderr, "No Flux handle found in publish_via_flux!\n");
+        return DYAD_NOCTX;
+    }
+    // TODO FLUX_LOG_INFO
+    printf("Create KVS transaction\n");
+    txn = flux_kvs_txn_create();
+    if (txn == NULL)
+    {
+        DYAD_LOG_ERR(ctx, "Could not create Flux KVS transaction\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Pack producer rank into transaction\n");
+    if (flux_kvs_txn_pack(txn, 0, topic, "i", ctx->rank) < 0)
+    {
+        DYAD_LOG_ERR(ctx, "Could not pack Flux KVS transaction\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Calling dyad_kvs_commit\n");
+    int ret = dyad_kvs_commit(ctx, txn, upath, &f);
+    if (ret < 0)
+    {
+        DYAD_LOG_ERR(ctx, "dyad_kvs_commit failed!\n");
+        return ret;
+    }
+    return DYAD_OK;
+}
+
+int dyad_commit(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = validate_context(ctx, false, true);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    if (ctx == NULL)
+    {
+        fprintf (stderr, "DYAD context not provided to dyad_commit\n");
+        return DYAD_NOCTX;
+    }
+    char upath[PATH_MAX] = {'\0'};
+    if (!cmp_canonical_path_prefix (ctx->prod_managed_path, fname, upath, PATH_MAX))
+    {
+        DYAD_LOG_INFO (ctx, "%s is not in the Producer's managed path\n", fname);
+        rc = 0;
+        goto commit_done;
+    }
+    ctx->reenter = false;
+    rc = publish_via_flux(ctx, upath);
+    ctx->reenter = true;
+
+commit_done:
+    if (rc == DYAD_OK && (ctx && ctx->check))
+    {
+        setenv(DYAD_CHECK_ENV, "ok", 1);
+    }
+    return rc;
+}
+
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int dyad_kvs_lookup(dyad_ctx_t *ctx, const char* kvs_topic, uint32_t *owner_rank)
```

Comment:
can ctx be const?

Reply:
There's two reasons I removed `__restrict__`:

1. It's not part of the C or C++ standards. I tried to minimize the amount of non-standard syntax in core to avoid any possible issues with using it in API libraries (e.g., stream and sync)
2. Although `__restrict__` is pretty well supported by most C compilers, most of what I'm seeing online says that support for `__restrict__` is much more iffy for C++ compilers. Since these functions are meant to be called from C++ (and possibly other languages in the future), I wanted to avoid `__restrict__` in the internal API for core.

That being said, functions like `dyad_kvs_lookup` that are not defined in the header could probably use `__restrict__`. I'll look into it for those.

Reply:
@JaeseungYeom if I were to add the restrict feature, would it be better to use `__restrict__` (i.e., compiler extension) or `restrict` (i.e., C99 keyword). Since our code is C99 (a requirement we inherit from Flux), we could just used the keyword, which should be supported by all C99-compliant compilers.